solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -0,0 +1,135 @@
+"""
+Provides preprocessing steps to be performed on the timeseries data.
+"""
+
+import pandas as pd
+
+from solarforecastarbiter import datamodel
+from solarforecastarbiter.validation import quality_mapping
+
+
+def apply_validation(data, qfilter, handle_func):
+    """
+    Apply validation steps based on provided filters to the data.
+
+    Parameters
+    ----------
+    qfilter : :class:`solarforecastarbiter.datamodel.QualityFlagFilter`
+    data : :class:`pd.DataFrame`
+        Pandas DataFrame of observation and forecast values.
+    handle_func : function
+        Function that handles how `quality_flags` will be used.
+        See :func:`solarforecastarbiter.metrics.preprocessing.exclude` as as
+        example.
+
+    Returns
+    -------
+    pd.DataFrame
+    """
+    validated_data = {}
+
+    # List of flags from filter
+    if not isinstance(qfilter, datamodel.QualityFlagFilter):
+        raise TypeError(f"{qfilter} not a QualityFlagFilter")
+    filters = qfilter.quality_flags
+
+    # Apply handling function to quality flags
+    for model, values in data.items():
+
+        # Skip if empty
+        if isinstance(values, pd.DataFrame) and values.empty:
+            validated_data[model] = values.value
+            continue
+
+        # Apply only to Observations
+        if isinstance(model, datamodel.Observation):
+            validation_df = quality_mapping.convert_mask_into_dataframe(
+                values['quality_flag'])
+            validation_df = validation_df[filters]
+            validated_data[model] = handle_func(values.value,
+                                                validation_df)
+        elif isinstance(model, datamodel.Forecast):
+            validated_data[model] = values
+        else:
+            raise TypeError(f"{model} not an Observation or Forecast")
+
+    return validated_data
+
+
+def resample_and_align(fx_obs, data, tz):
+    """
+    Resample the observation to the forecast interval length and align to
+    remove overlap.
+
+    Parameters
+    ----------
+    fx_obs : :class:`solarforecastarbiter.datamodel.ForecastObservation`
+        Pair of forecast and observation.
+    data : dict
+        Dictionary that must include the timeseries of the pair and already
+        be validated.
+    tz : str
+        Timezone to witch processed data will be converted.
+
+    Returns
+    -------
+    :class:`solarforecastarbiter.datamodel.ProcessedForecastObservation`
+
+    Todo
+    ----
+      * Add other resampling functions (besides mean like first, last, median)
+    """
+    fx = fx_obs.forecast
+    obs = fx_obs.observation
+
+    # Resample observation
+    closed = datamodel.CLOSED_MAPPING[fx.interval_label]
+    obs_resampled = data[obs].resample(fx.interval_length,
```

Comment:
Maybe out of scope here, but this will need to change when we want to account for quality flags during resampling. Not sure if that's going to require a change to the `ProcessedForecastObservation` too.

Reply:
My current thinking (it has changed often) is that the quality flags will be handled independently from the resampling.  Of course resampling does allow for fillin itself. Perhaps the `apply_validation` should only "compute" the boolean of the quality flag but as you pointed out then ProcessedForecastObservation would need a DataFrame for the observation_value or observation_quality_flag attribute.